crates/sui-core/src/epoch/consensus_store_pruner.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use mysten_metrics::spawn_logged_monitored_task;
+use narwhal_config::Epoch;
+use std::fs;
+use std::path::PathBuf;
+use std::time::Duration;
+use tokio::{sync::mpsc, time::Instant};
+use tracing::{error, info};
+use typed_store::rocks::safe_drop_db;
+
+pub struct ConsensusStorePruner {
+    base_path: PathBuf,
+    tx_remove: mpsc::Sender<Epoch>,
+    epoch_retention: u64,
+    epoch_prune_period: Duration,
+}
+
+impl ConsensusStorePruner {
+    pub fn new(base_path: PathBuf, epoch_retention: u64, epoch_prune_period: Duration) -> Self {
+        let (tx_remove, _rx_remove) = mpsc::channel(1);
+        Self {
+            base_path,
+            tx_remove,
+            epoch_retention,
+            epoch_prune_period,
+        }
+    }
+
+    pub async fn run(&mut self) {
+        let (tx_remove, mut rx_remove) = mpsc::channel(1);
+        self.tx_remove = tx_remove;
+        let base_path = self.base_path.clone();
+        let epoch_retention = self.epoch_retention;
+        let epoch_prune_period = self.epoch_prune_period;
+        let mut latest_epoch = 0;
+
+        spawn_logged_monitored_task!(async {
+            info!("Starting consensus store pruner with epoch retention {epoch_retention} and prune period {epoch_prune_period:?}");
+
+            let mut timeout = tokio::time::interval_at(
+                Instant::now() + Duration::from_secs(60), // allow some time for the node to boot etc before attempting to prune
+                epoch_prune_period,
+            );
+            loop {
+                tokio::select! {
+                    _ = timeout.tick() => {
+                        Self::prune_old_epoch_data(base_path.clone(), latest_epoch, epoch_retention).await;
+                    }
+                    result = rx_remove.recv() => {
+                        if result.is_none() {
+                            info!("Closing consensus store pruner");
+                            break;
+                        }
+                        latest_epoch = result.unwrap();
+                        Self::prune_old_epoch_data(base_path.clone(), latest_epoch, epoch_retention).await;
+                    }
+                }
+            }
+        });
+    }
+
+    /// This method will remove all epoch data stores and directories that are older than the current epoch minus the epoch retention. The method ensures
+    /// that always the `current_epoch` data is retained.
+    pub async fn prune(&self, current_epoch: Epoch) {
+        let result = self.tx_remove.send(current_epoch).await;
+        if result.is_err() {
+            error!(
+                "Error sending message to data removal task for epoch {:?}",
+                current_epoch,
+            );
+        }
+    }
+
+    pub(crate) async fn prune_old_epoch_data(
```

Comment:
Why `pub(crate)`?

Reply:
please ignore - copy/paste from old code. Before the unit test was in separate file and I guess that might needed this visibility